core/plugins/org.polarsys.capella.core.data.interaction.validation/src/org/polarsys/capella/core/data/interaction/validation/sequenceMessage/SequenceMessageInvokedOperationExchangeItems.java

```diff
@@ -0,0 +1,98 @@
+import org.eclipse.emf.validation.AbstractModelConstraint;
+import org.eclipse.emf.validation.EMFEventType;
+import org.eclipse.emf.validation.IValidationContext;
+import org.polarsys.capella.common.data.modellingcore.AbstractExchangeItem;
+import org.polarsys.capella.common.helpers.EObjectLabelProviderHelper;
+import org.polarsys.capella.core.data.cs.ExchangeItemAllocation;
+import org.polarsys.capella.core.data.fa.AbstractFunction;
+import org.polarsys.capella.core.data.fa.ComponentExchange;
+import org.polarsys.capella.core.data.fa.FunctionalExchange;
+import org.polarsys.capella.core.data.information.AbstractEventOperation;
+import org.polarsys.capella.core.data.information.Operation;
+import org.polarsys.capella.core.data.information.Service;
+import org.polarsys.capella.core.data.interaction.MessageKind;
+import org.polarsys.capella.core.data.interaction.SequenceMessage;
+import org.polarsys.capella.core.data.interaction.properties.dialogs.sequenceMessage.model.communications.InterfaceCommunication;
+import org.polarsys.capella.core.data.oa.CommunicationMean;
+import org.polarsys.capella.core.data.oa.OperationalActivity;
+import org.polarsys.capella.core.model.helpers.FunctionalExchangeExt;
+import org.polarsys.capella.core.model.helpers.SequenceMessageExt;
+
+/**
+ */
+public class SequenceMessageInvokedOperationExchangeItems extends AbstractModelConstraint {
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public IStatus validate(IValidationContext ctx) {
+        //
+        // Preconditions.
+        //
+        EMFEventType eType = ctx.getEventType();
+        if (EMFEventType.NULL != eType) {
+            return ctx.createSuccessStatus();
+        }
+        EObject eObj = ctx.getTarget();
+        if (!(eObj instanceof SequenceMessage)) {
+            return ctx.createSuccessStatus();
+        }
+        SequenceMessage sequenceMessage = (SequenceMessage) eObj;
+        // Ignore TIMER SequenceMessage (by construction, they don't have a invoked operation).
+        if (MessageKind.TIMER == sequenceMessage.getKind()) {
+            return ctx.createSuccessStatus();
+        }
+        AbstractEventOperation invokedOperation = sequenceMessage.getInvokedOperation();
+        if (null != invokedOperation) {
+            Collection<AbstractExchangeItem> lstInvokedEI = SequenceMessageExt.getExchangeItemsFromOperation(sequenceMessage);
+            if (!sequenceMessage.getExchangedItems().containsAll(lstInvokedEI)) {
+                String invokedOperationDisplayName = computeDisplayNameForOperation(invokedOperation);
+                return ctx.createFailureStatus(sequenceMessage.getName(), invokedOperation.getName(), invokedOperationDisplayName);
+            }
+        }
+        return ctx.createSuccessStatus();
+    }
+    
+    private String computeDisplayNameForOperation(AbstractEventOperation invokedOperation) {
+        String result = "";
+        if (invokedOperation instanceof CommunicationMean) {
```

Comment:
Could you refactor the method? There are a lot of calls to getMetaclassLabel. Is there any AbstractEventOperation type that shouldn't be included in the processing?

Reply:
You are right, there is a special treatment only for "Interaction" Exchanges but covering every concrete subtype of AbstractEventOperation ensures to have no surprise in the future.